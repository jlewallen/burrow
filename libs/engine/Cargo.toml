[package]
name = "engine"
version = "0.1.0"
authors = ["Jacob Lewallen <jlewallen@gmail.com>"]
edition = "2021"

[lib]
doctest = false

[dependencies.kernel]
path = "../kernel"

[dependencies]
anyhow = "1.0.72"
chrono = { version = "0.4.26", features = ["serde"] }
serde = { version = "1.0.180", features = ["derive"] }
serde_json = "1.0.104"
tracing = "0.1.37"
tracing-subscriber = { version = "0.3.17", features = ["json"] }
