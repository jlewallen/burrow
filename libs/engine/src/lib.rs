mod identifiers;
mod internal;
mod perform;
mod users;

pub mod domain;
pub mod notifications;
pub mod sequences;
pub mod session;
pub mod storage;

pub use domain::*;
pub use notifications::*;
pub use session::*;

pub use users::model::{add_username_to_key, username_to_key};
