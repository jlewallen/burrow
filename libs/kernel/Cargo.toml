[package]
name = "kernel"
version = "0.1.0"
authors = ["Jacob Lewallen <jlewallen@gmail.com>"]
edition = "2021"

[lib]
doctest = false

[dependencies.replies]
path = "../replies"

[dependencies]
anyhow = "1.0.71"
thiserror = "1.0.40"
tracing = "0.1.37"
nom = "7.1.3"
serde_json = "1.0.96"
serde = { version = "1.0.163", features = ["derive"] }
treediff = { version = "4.0.2", features = ["with-serde-json"] }