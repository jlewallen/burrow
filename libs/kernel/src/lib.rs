pub mod actions;
pub mod hooks;
pub mod model;
pub mod perms;
pub mod plugins;
pub mod session;
pub mod surround;

pub mod common {
    pub use replies::*;
}

pub mod prelude {
    pub use replies::DomainEvent;

    pub use crate::actions::*;
    pub use crate::finder::*;
    pub use crate::hooks::*;
    pub use crate::model::compare::*;
    pub use crate::model::*;
    pub use crate::plugins::*;
    pub use crate::session::*;
    pub use crate::surround::*;
}

mod finder {
    use crate::model::{Audience, EntityKey, Entry, Item};
    use crate::surround::Surroundings;

    pub trait Finder: Send + Sync {
        fn find_world(&self) -> anyhow::Result<Entry>;

        fn find_location(&self, entry: &Entry) -> anyhow::Result<Entry>;

        fn find_item(
            &self,
            surroundings: &Surroundings,
            item: &Item,
        ) -> anyhow::Result<Option<Entry>>;

        fn find_audience(&self, audience: &Audience) -> anyhow::Result<Vec<EntityKey>>;
    }
}

#[macro_export]
macro_rules! here {
    () => {{
        format!("{}:{}", file!(), line!())
    }};
}

#[cfg(test)]
mod tests;
