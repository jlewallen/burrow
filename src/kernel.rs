use anyhow::Result;
use markdown_gen::markdown;
use once_cell::sync::Lazy;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::{collections::HashMap, fmt::Display, ops::Index, rc::Rc, string::FromUtf8Error};
use thiserror::Error;
use tracing::{debug, span, Level};

pub static WORLD_KEY: Lazy<EntityKey> = Lazy::new(|| "world".to_string());

pub type EntityKey = String;

pub type ActionArgs<'a> = (&'a Entity, &'a Entity, &'a Entity);

pub type Markdown = markdown::Markdown<Vec<u8>>;

pub fn markdown_to_string(md: Markdown) -> Result<String, FromUtf8Error> {
    String::from_utf8(md.into_inner())
}

type BoxedScope<T> = Box<T>;

pub trait DomainEvent: std::fmt::Debug {}

#[derive(Debug)]
pub struct DomainResult {
    pub events: Vec<Box<dyn DomainEvent>>,
}

pub trait Reply: std::fmt::Debug + erased_serde::Serialize {
    fn to_markdown(&self) -> Result<Markdown>;
}

pub trait InfrastructureFactory: std::fmt::Debug {
    fn new_infrastructure(&self) -> Result<Rc<dyn DomainInfrastructure>>;
}

pub trait DomainInfrastructure: std::fmt::Debug {
    fn ensure_entity(&self, entity_ref: &DynamicEntityRef) -> Result<DynamicEntityRef>;

    fn ensure_optional_entity(
        &self,
        entity_ref: &Option<DynamicEntityRef>,
    ) -> Result<Option<DynamicEntityRef>> {
        match entity_ref {
            Some(e) => Ok(Some(self.ensure_entity(&e)?)),
            None => Ok(None),
        }
    }
}

pub trait PrepareWithInfrastructure {
    fn prepare_with(&mut self, infra: &dyn DomainInfrastructure) -> Result<()>;
}

pub trait Action: std::fmt::Debug {
    fn perform(&self, args: ActionArgs) -> Result<Box<dyn Reply>>;
}

pub trait Scope: PrepareWithInfrastructure + DeserializeOwned {
    fn scope_key() -> &'static str
    where
        Self: Sized;
}

pub trait PrepareEntityByKey {
    fn prepare_entity_by_key<T: Fn(&mut Entity) -> Result<()>>(
        &self,
        key: &EntityKey,
        prepare: T,
    ) -> Result<&Entity>;
}

pub trait LoadEntityByKey {
    fn load_entity_by_key(&self, key: &EntityKey) -> Result<&Entity>;

    fn load_entity_by_ref(&self, entity_ref: &EntityRef) -> Result<&Entity> {
        self.load_entity_by_key(&entity_ref.key)
    }

    fn load_entities_by_refs(&self, entity_refs: Vec<EntityRef>) -> Result<Vec<&Entity>> {
        entity_refs
            .into_iter()
            .map(|re| -> Result<&Entity> { self.load_entity_by_ref(&re) })
            .collect()
    }

    fn load_entities_by_keys(&self, entity_keys: Vec<EntityKey>) -> Result<Vec<&Entity>> {
        entity_keys
            .into_iter()
            .map(|key| -> Result<&Entity> { self.load_entity_by_key(&key) })
            .collect()
    }
}

#[derive(Debug, Serialize)]
pub enum SimpleReply {
    Done,
}

impl Reply for SimpleReply {
    fn to_markdown(&self) -> Result<Markdown> {
        let mut md = Markdown::new(Vec::new());
        md.write("ok!")?;
        Ok(md)
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Item {
    Named(String),
    // ImplicitlyUnheld(String),
    // ImplicitlyHeld(String),
    // ImplicitlyNavigable(String),
}

#[derive(Error, Debug)]
pub enum DomainError {
    #[error("no such scope '{:?}' on entity '{:?}'", .0, .1)]
    NoSuchScope(EntityKey, String),
    #[error("parse failed")]
    ParseFailed(#[source] serde_json::Error),
    #[error("dangling entity")]
    DanglingEntity,
    #[error("anyhow")]
    Anyhow(#[source] anyhow::Error),
    #[error("no infrastructure")]
    NoInfrastructure,
}

impl From<serde_json::Error> for DomainError {
    fn from(source: serde_json::Error) -> Self {
        DomainError::ParseFailed(source)
    }
}

impl From<anyhow::Error> for DomainError {
    fn from(source: anyhow::Error) -> Self {
        DomainError::Anyhow(source)
    }
}

#[derive(Error, Debug)]
pub enum EvaluationError {
    #[error("parse failed")]
    ParseFailed,
}

impl From<nom::Err<nom::error::Error<&str>>> for EvaluationError {
    fn from(source: nom::Err<nom::error::Error<&str>>) -> EvaluationError {
        match source {
            nom::Err::Incomplete(_) => EvaluationError::ParseFailed,
            nom::Err::Error(_e) => EvaluationError::ParseFailed,
            nom::Err::Failure(_e) => EvaluationError::ParseFailed,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityRef {
    #[serde(alias = "py/object")]
    py_object: String,
    #[serde(alias = "py/ref")]
    py_ref: String,
    key: EntityKey,
    klass: String,
    name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Identity {
    #[serde(alias = "py/object")]
    py_object: String,
    private: String,
    public: String,
    signature: Option<String>, // TODO Why does this happen in the model?
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Kind {
    #[serde(alias = "py/object")]
    py_object: String,
    identity: Identity,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityClass {
    #[serde(alias = "py/type")]
    py_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AclRule {
    #[serde(alias = "py/object")]
    py_object: String,
    keys: Vec<String>,
    perm: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Acls {
    #[serde(alias = "py/object")]
    py_object: String,
    rules: Vec<AclRule>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Version {
    #[serde(alias = "py/object")]
    py_object: String,
    i: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Property {
    value: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Props {
    map: HashMap<String, Property>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Entity {
    #[serde(alias = "py/object")]
    py_object: String,
    pub key: String,
    version: Version,
    parent: Option<EntityRef>,
    creator: Option<EntityRef>,
    identity: Identity,
    #[serde(alias = "klass")]
    class: EntityClass,
    acls: Acls,
    props: Props,
    scopes: HashMap<String, serde_json::Value>,

    #[serde(skip)] // Very private
    infra: Option<Rc<dyn DomainInfrastructure>>,
}

impl Display for Entity {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        fmt.debug_struct(&self.class.py_type)
            .field("key", &self.key)
            .field("name", &self.name())
            .finish()
    }
}

impl Entity {
    pub fn set_infra(&mut self, infra: Rc<dyn DomainInfrastructure>) {
        self.infra = Some(infra);
    }

    fn property_named(&self, name: &str) -> Option<&Property> {
        if self.props.map.contains_key(name) {
            return Some(self.props.map.index(name));
        }
        None
    }

    pub fn name(&self) -> Option<String> {
        if let Some(property) = self.property_named("name") {
            match &property.value {
                serde_json::Value::String(v) => Some(v.to_string()),
                _ => None,
            }
        } else {
            None
        }
    }

    pub fn has_scope<T: Scope>(&self) -> bool {
        self.scopes.contains_key(<T as Scope>::scope_key())
    }

    pub fn scope<T: Scope>(&self) -> Result<BoxedScope<T>, DomainError> {
        let scope_key = <T as Scope>::scope_key();

        let _load_scope_span =
            span!(Level::DEBUG, "scope", key = self.key, scope = scope_key).entered();

        if !self.scopes.contains_key(scope_key) {
            return Err(DomainError::NoSuchScope(
                self.key.clone(),
                scope_key.to_string(),
            ));
        }

        let data = &self.scopes[scope_key];

        debug!("parsing");

        // The call to serde_json::from_value requires owned data and we have a
        // reference to somebody else's. Presumuably so that we don't couple the
        // lifetime of the returned object to the lifetime of the data being
        // referenced? What's the right solution here?
        // Should the 'un-parsed' Scope also owned the parsed data?
        let owned_value = data.clone();
        let mut scope: Box<T> = serde_json::from_value(owned_value)?;

        let _prepare_span = span!(Level::DEBUG, "prepare").entered();

        if let Some(infra) = &self.infra {
            scope.prepare_with(infra.as_ref())?;
        } else {
            return Err(DomainError::NoInfrastructure);
        }

        // Ok, great!
        Ok(scope)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum DynamicEntityRef {
    RefOnly {
        #[serde(alias = "py/object")]
        py_object: String,
        #[serde(alias = "py/ref")]
        py_ref: String,
        key: String,
        klass: String,
        name: String,
    },
    Entity(Box<Entity>),
}

impl DynamicEntityRef {
    pub fn key(&self) -> &String {
        match self {
            DynamicEntityRef::RefOnly {
                py_object: _,
                py_ref: _,
                key,
                klass: _,
                name: _,
            } => key,
            DynamicEntityRef::Entity(e) => &e.key,
        }
    }
}

impl TryFrom<DynamicEntityRef> for Box<Entity> {
    type Error = DomainError;

    fn try_from(value: DynamicEntityRef) -> Result<Self, Self::Error> {
        match value {
            DynamicEntityRef::RefOnly {
                py_object: _,
                py_ref: _,
                key: _,
                klass: _,
                name: _,
            } => Err(DomainError::DanglingEntity),
            DynamicEntityRef::Entity(e) => Ok(e),
        }
    }
}

impl From<DynamicEntityRef> for EntityRef {
    fn from(value: DynamicEntityRef) -> Self {
        match value {
            DynamicEntityRef::RefOnly {
                py_object,
                py_ref,
                key,
                klass,
                name,
            } => EntityRef {
                py_object,
                py_ref,
                key,
                klass,
                name,
            },
            DynamicEntityRef::Entity(e) => e.as_ref().into(),
        }
    }
}

impl From<&Entity> for EntityRef {
    fn from(e: &Entity) -> Self {
        EntityRef {
            py_object: "todo!".to_string(),
            py_ref: "todo!".to_string(),
            key: e.key.to_string(),
            klass: "todo!".to_string(),
            name: e.name().unwrap_or_default(),
        }
    }
}

#[derive(Debug)]
pub struct PersistedEntity {
    pub key: String,
    pub gid: u32,
    pub version: u32,
    pub serialized: String,
}
