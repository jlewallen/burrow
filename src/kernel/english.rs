use nom::combinator::opt;
pub use nom::{
    branch::alt,
    bytes::complete::{tag, take_while, take_while1},
    character::complete::digit1,
    combinator::map,
    combinator::{map_res, recognize},
    multi::separated_list0,
    sequence::{delimited, pair, preceded, separated_pair, terminated, tuple},
    IResult,
};
pub use tracing::*;

#[derive(Debug, Clone, PartialEq)]
pub enum English {
    Literal(String),
    Phrase(Box<Vec<English>>),
    OneOf(Box<Vec<English>>),
    Optional(Box<English>),
    Unheld,
    Held,
    Contained,
    Numbered(u64),
}

pub fn to_english(i: &str) -> IResult<&str, Vec<English>> {
    separated_list0(spaces, term)(i)
}

fn term(i: &str) -> IResult<&str, English> {
    alt((literal, contained, unheld, held, numbered, optional_phrase))(i)
}

fn optional_phrase(i: &str) -> IResult<&str, English> {
    map(
        tuple((phrase, opt(tag("?")))),
        |(p, is_optional)| match is_optional {
            Some(_) => English::Optional(Box::new(p)),
            None => p,
        },
    )(i)
}

fn phrase(i: &str) -> IResult<&str, English> {
    map(
        delimited(
            tag("("),
            separated_list0(tag("|"), separated_list0(spaces, term)),
            tag(")"),
        ),
        |optionals| {
            if optionals.len() == 1 {
                match optionals.get(0) {
                    Some(v) => English::Phrase(Box::new(v.to_vec())),
                    None => todo!(),
                }
            } else {
                English::OneOf(Box::new(
                    optionals
                        .into_iter()
                        .map(|e| English::Phrase(Box::new(e)))
                        .collect(),
                ))
            }
        },
    )(i)
}

fn spaces(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| " \t".contains(c))(i)
}

fn unsigned_number(i: &str) -> IResult<&str, u64> {
    map_res(recognize(digit1), str::parse)(i)
}

fn numbered(i: &str) -> IResult<&str, English> {
    map(preceded(tag("#"), unsigned_number), |n| {
        English::Numbered(n)
    })(i)
}

fn unheld(i: &str) -> IResult<&str, English> {
    map(tag("#unheld"), |_| English::Unheld)(i)
}

fn held(i: &str) -> IResult<&str, English> {
    map(tag("#held"), |_| English::Held)(i)
}

fn contained(i: &str) -> IResult<&str, English> {
    map(tag("#contained"), |_| English::Contained)(i)
}

fn uppercase_word(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| "ABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
}

fn literal(i: &str) -> IResult<&str, English> {
    map(uppercase_word, |v| English::Literal(v.into()))(i)
}

#[cfg(test)]
mod tests {
    use super::*;

    struct Fixture {
        text: &'static str,
        expected: Vec<English>,
    }

    impl Fixture {
        pub fn new(text: &'static str, expected: Vec<English>) -> Self {
            Self { text, expected }
        }
    }

    fn get_fixtures() -> Vec<Fixture> {
        vec![
            Fixture::new(
                "HOLD #unheld",
                vec![English::Literal("HOLD".into()), English::Unheld],
            ),
            Fixture::new(
                "PUT #held IN #held",
                vec![
                    English::Literal("PUT".into()),
                    English::Held,
                    English::Literal("IN".into()),
                    English::Held,
                ],
            ),
            Fixture::new(
                "PUT #held (INSIDE OF|IN) (#held|#unheld)?",
                vec![
                    English::Literal("PUT".into()),
                    English::Held,
                    English::OneOf(Box::new(vec![
                        English::Phrase(Box::new(vec![
                            English::Literal("INSIDE".into()),
                            English::Literal("OF".into()),
                        ])),
                        English::Phrase(Box::new(vec![English::Literal("IN".into())])),
                    ])),
                    English::Optional(Box::new(English::OneOf(Box::new(vec![
                        English::Phrase(Box::new(vec![English::Held])),
                        English::Phrase(Box::new(vec![English::Unheld])),
                    ])))),
                ],
            ),
            Fixture::new(
                "TAKE (OUT)? #contained (OUT OF (#held|#unheld))?",
                vec![
                    English::Literal("TAKE".into()),
                    English::Optional(Box::new(English::Phrase(Box::new(vec![English::Literal(
                        "OUT".into(),
                    )])))),
                    English::Contained,
                    English::Optional(Box::new(English::Phrase(Box::new(vec![
                        English::Literal("OUT".into()),
                        English::Literal("OF".into()),
                        English::OneOf(Box::new(vec![
                            English::Phrase(Box::new(vec![English::Held])),
                            English::Phrase(Box::new(vec![English::Unheld])),
                        ])),
                    ])))),
                ],
            ),
            Fixture::new(
                "HOLD #unheld",
                vec![English::Literal("HOLD".into()), English::Unheld],
            ),
            Fixture::new(
                "DROP (#held)?",
                vec![
                    English::Literal("DROP".into()),
                    English::Optional(Box::new(English::Phrase(Box::new(vec![English::Held])))),
                ],
            ),
            Fixture::new(
                "EDIT #3493",
                vec![English::Literal("EDIT".into()), English::Numbered(3493)],
            ),
        ]
    }

    #[test]
    fn should_parse_all_english_fixtures() {
        for fixture in get_fixtures() {
            let (remaining, actual) = to_english(fixture.text).unwrap();
            assert_eq!(remaining, "");
            assert_eq!(actual, fixture.expected);
        }
    }
}
