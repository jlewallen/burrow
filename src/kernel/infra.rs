use anyhow::Result;
use std::{cell::RefCell, rc::Rc};

use super::{
    Action, ActionArgs, DomainError, EntityGID, EntityKey, EntityPtr, EntityRef, Item,
    LazyLoadedEntity, Reply,
};

thread_local! {
    #[allow(unused)]
    static SESSION: RefCell<Option<std::rc::Weak<dyn Infrastructure>>> = RefCell::new(None)
}

pub fn set_my_session(session: Option<&Rc<dyn Infrastructure>>) -> Result<()> {
    SESSION.with(|s| {
        *s.borrow_mut() = match session {
            Some(session) => Some(Rc::downgrade(session)),
            None => None,
        };

        Ok(())
    })
}

pub fn get_my_session() -> Result<Rc<dyn Infrastructure>> {
    SESSION.with(|s| match &*s.borrow() {
        Some(s) => match s.upgrade() {
            Some(s) => Ok(s),
            None => Err(DomainError::ExpiredInfrastructure.into()),
        },
        None => Err(DomainError::NoInfrastructure.into()),
    })
}

pub trait LoadEntities {
    fn load_entity_by_key(&self, key: &EntityKey) -> Result<Option<EntityPtr>>;

    fn load_entity_by_gid(&self, gid: &EntityGID) -> Result<Option<EntityPtr>>;

    fn load_entity_by_ref(&self, entity_ref: &EntityRef) -> Result<Option<EntityPtr>> {
        self.load_entity_by_key(&entity_ref.key)
    }
}

/// I think this will eventually need to return or take a construct that's
/// builder-like so callers can take more control. Things to consider are:
/// 1) Conditional needle visibility.
/// 2) Items containing others.
/// 3) Verb capabilities of the needle.
pub trait FindsItems {
    fn find_item(&self, args: ActionArgs, item: &Item) -> Result<Option<EntityPtr>>;

    fn find_optional_item(
        &self,
        args: ActionArgs,
        item: Option<Item>,
    ) -> Result<Option<EntityPtr>> {
        if let Some(item) = item {
            self.find_item(args, &item)
        } else {
            Ok(None)
        }
    }
}

pub trait Infrastructure: LoadEntities + FindsItems {
    fn ensure_entity(&self, entity_ref: &LazyLoadedEntity) -> Result<LazyLoadedEntity>;

    fn ensure_optional_entity(
        &self,
        entity_ref: &Option<LazyLoadedEntity>,
    ) -> Result<Option<LazyLoadedEntity>> {
        match entity_ref {
            Some(e) => Ok(Some(self.ensure_entity(e)?)),
            None => Ok(None),
        }
    }

    fn add_entity(&self, entity: &EntityPtr) -> Result<()>;

    fn add_entities(&self, entities: &Vec<&EntityPtr>) -> Result<()> {
        for entity in entities {
            self.add_entity(entity)?;
        }
        Ok(())
    }

    fn chain(&self, living: &EntityPtr, action: Box<dyn Action>) -> Result<Box<dyn Reply>>;
}

pub trait Needs<T> {
    fn supply(&mut self, resource: &T) -> Result<()>;
}

pub trait SessionTrait: Infrastructure {}
