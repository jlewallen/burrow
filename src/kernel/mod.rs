pub mod english;
pub mod infra;
pub mod model;
pub mod scopes;

pub use english::*;
pub use infra::*;
pub use model::*;
pub use scopes::*;

pub use replies::*;

pub type ReplyResult = anyhow::Result<Box<dyn Reply>>;
