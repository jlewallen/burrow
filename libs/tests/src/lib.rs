#[cfg(test)]
mod tests {
    use std::{rc::Rc, sync::Arc};

    use anyhow::Result;
    use engine::{
        storage, DevNullNotifier, Domain, EntityStorageFactory, Finder, PersistedEntity, Session,
        SessionOpener,
    };
    use kernel::RegisteredPlugins;
    use plugins_core::{
        building::BuildingPluginFactory, carrying::CarryingPluginFactory,
        dynamic::DynamicPluginFactory, looking::LookingPluginFactory, moving::MovingPluginFactory,
        BuildSurroundings, DefaultFinder, QuickThing,
    };
    use plugins_rpc::RpcPluginFactory;
    use plugins_rune::RunePluginFactory;
    use plugins_wasm::WasmPluginFactory;
    use tokio::{runtime::Handle, task::JoinHandle};

    const USERNAME: &str = "burrow";

    #[derive(Clone)]
    struct AsyncFriendlyDomain {
        domain: Domain,
    }

    impl AsyncFriendlyDomain {
        pub fn new<SF>(
            storage_factory: Arc<SF>,
            plugins: Arc<RegisteredPlugins>,
            finder: Arc<dyn Finder>,
            deterministic: bool,
        ) -> Self
        where
            SF: EntityStorageFactory + 'static,
        {
            Self {
                domain: Domain::new(storage_factory, plugins, finder, deterministic),
            }
        }

        pub async fn query_all(&self) -> Result<Vec<PersistedEntity>> {
            self.domain.query_all()
        }

        #[cfg(test)]
        pub async fn snapshot(&self) -> Result<serde_json::Value> {
            let json: Vec<serde_json::Value> = self
                .query_all()
                .await?
                .into_iter()
                .map(|p| p.to_json_value())
                .collect::<Result<_>>()?;

            Ok(serde_json::Value::Array(json))
        }

        pub async fn evaluate<W>(&self, text: &'static [&'static str]) -> Result<()>
        where
            W: WorldFixture + Default,
        {
            let handle: JoinHandle<Result<()>> = tokio::task::spawn_blocking({
                let sessions = self.clone();
                move || {
                    let session = sessions.open_session()?;

                    let fixture = W::default();

                    fixture.prepare(&session)?;

                    for text in text {
                        if let Some(reply) = session.evaluate_and_perform(USERNAME, text)? {
                            println!("{:?}", &reply);
                        }
                    }

                    session.close(&DevNullNotifier::default())?;

                    Ok(())
                }
            });

            Ok(handle.await??)
        }

        pub async fn stop(&self) -> Result<()> {
            let domain = self.domain.clone();
            tokio::task::spawn_blocking(move || domain.stop()).await?
        }
    }

    impl SessionOpener for AsyncFriendlyDomain {
        fn open_session(&self) -> Result<std::rc::Rc<Session>> {
            self.domain.open_session()
        }
    }

    async fn test_domain() -> Result<AsyncFriendlyDomain> {
        let storage_factory = storage::sqlite::Factory::new(storage::sqlite::MEMORY_SPECIAL)?;
        let mut registered_plugins = RegisteredPlugins::default();
        registered_plugins.register(MovingPluginFactory::default());
        registered_plugins.register(LookingPluginFactory::default());
        registered_plugins.register(CarryingPluginFactory::default());
        registered_plugins.register(BuildingPluginFactory::default());
        registered_plugins.register(DynamicPluginFactory::default());
        registered_plugins.register(RunePluginFactory::default());
        registered_plugins.register(WasmPluginFactory::default());
        registered_plugins.register(RpcPluginFactory::start(Handle::current()).await?);
        let finder = Arc::new(DefaultFinder::default());
        Ok(AsyncFriendlyDomain::new(
            storage_factory,
            Arc::new(registered_plugins),
            finder,
            true,
        ))
    }

    trait WorldFixture {
        fn prepare(&self, session: &Rc<Session>) -> Result<()>;
    }

    #[derive(Default)]
    struct Noop {}

    impl WorldFixture for Noop {
        fn prepare(&self, _session: &Rc<Session>) -> Result<()> {
            Ok(())
        }
    }

    #[derive(Default)]
    struct KeyInVessel {}

    impl WorldFixture for KeyInVessel {
        fn prepare(&self, session: &Rc<Session>) -> Result<()> {
            let mut build = BuildSurroundings::new_in_session(session.clone())?;
            let key = build.entity()?.named("Key")?.into_entry()?;
            let vessel = build
                .entity()?
                .named("Vessel")?
                .holding(&vec![key.clone()])?
                .into_entry()?;
            let (_session, _surroundings) = build
                .hands(vec![QuickThing::Actual(vessel.clone())])
                .build()?;

            session.flush()?;

            Ok(())
        }
    }

    #[tokio::test]
    async fn it_evaluates_a_simple_look() -> Result<()> {
        let domain = test_domain().await?;
        domain.evaluate::<KeyInVessel>(&["look"]).await?;
        insta::assert_json_snapshot!(domain.snapshot().await?);
        domain.stop().await?;

        Ok(())
    }

    #[tokio::test]
    async fn it_evaluates_two_simple_looks_same_session() -> Result<()> {
        let domain = test_domain().await?;
        domain.evaluate::<KeyInVessel>(&["look", "look"]).await?;
        insta::assert_json_snapshot!(domain.snapshot().await?);
        domain.stop().await?;

        Ok(())
    }

    #[tokio::test]
    async fn it_evaluates_two_simple_looks_separate_session() -> Result<()> {
        let domain = test_domain().await?;
        domain.evaluate::<KeyInVessel>(&["look"]).await?;
        domain.evaluate::<Noop>(&["look"]).await?;
        insta::assert_json_snapshot!(domain.snapshot().await?);
        domain.stop().await?;

        Ok(())
    }
}

#[cfg(test)]
#[ctor::ctor]
fn initialize_tests() {
    plugins_core::log_test();
}
