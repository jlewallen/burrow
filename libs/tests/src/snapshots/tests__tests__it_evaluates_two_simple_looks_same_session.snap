---
source: libs/tests/src/lib.rs
expression: domain.snapshot().await?
---
[
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-0",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 1
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Key"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "location": {
        "container": {
          "gid": 2,
          "key": "E-1",
          "klass": "",
          "name": "Vessel",
          "py/object": "model.entity.EntityRef",
          "py/ref": "model.entity.Entity"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-1",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 2
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Vessel"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "containing": {
        "capacity": null,
        "holding": [
          {
            "gid": 1,
            "key": "E-0",
            "klass": "",
            "name": "Key",
            "py/object": "model.entity.EntityRef",
            "py/ref": "model.entity.Entity"
          }
        ],
        "produces": {}
      },
      "location": {
        "container": {
          "gid": 3,
          "key": "E-2",
          "klass": "",
          "name": "Living",
          "py/object": "model.entity.EntityRef",
          "py/ref": "model.entity.Entity"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-2",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 3
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Living"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "containing": {
        "capacity": null,
        "holding": [
          {
            "gid": 2,
            "key": "E-1",
            "klass": "",
            "name": "Vessel",
            "py/object": "model.entity.EntityRef",
            "py/ref": "model.entity.Entity"
          }
        ],
        "produces": {}
      },
      "occupying": {
        "area": {
          "gid": 5,
          "key": "E-4",
          "klass": "",
          "name": "Welcome Area",
          "py/object": "model.entity.EntityRef",
          "py/ref": "model.entity.Entity"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "world",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 4
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "World"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "identifiers": {
        "acls": {
          "py/object": "",
          "rules": []
        },
        "gid": 5
      },
      "usernames": {
        "users": {
          "burrow": "E-2"
        }
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  },
  {
    "acls": {
      "py/object": "",
      "rules": []
    },
    "creator": null,
    "identity": {
      "private": "",
      "public": "",
      "py/object": "",
      "signature": null
    },
    "key": "E-4",
    "klass": {
      "py/type": ""
    },
    "parent": null,
    "props": {
      "map": {
        "gid": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": 5
        },
        "name": {
          "acls": {
            "py/object": "",
            "rules": []
          },
          "py/object": "",
          "value": "Welcome Area"
        }
      },
      "py/object": "model.properties.Common"
    },
    "py/object": "",
    "scopes": {
      "containing": {
        "capacity": null,
        "holding": [],
        "produces": {}
      },
      "occupyable": {
        "acls": {
          "py/object": "",
          "rules": []
        },
        "occupancy": 0,
        "occupied": [
          {
            "gid": 3,
            "key": "E-2",
            "klass": "",
            "name": "Living",
            "py/object": "model.entity.EntityRef",
            "py/ref": "model.entity.Entity"
          }
        ]
      }
    },
    "version": {
      "i": 3,
      "py/object": ""
    }
  }
]
